pkg/appgateserver/sdkadapter/sdk.go

```diff
@@ -1,65 +1,141 @@
+	sessionClient, err := query.NewSessionQuerier(deps)
 	if err != nil {
 		return nil, err
 	}
 
-	sessionClient, err := query.NewSessionQuerier(deps)
+	accountClient, err := query.NewAccountQuerier(deps)
 	if err != nil {
 		return nil, err
 	}
 
-	accountClient, err := query.NewAccountQuerier(deps)
+	appClient, err := query.NewApplicationQuerier(deps)
 	if err != nil {
 		return nil, err
 	}
 
-	sharedParamsClient, err := query.NewSharedQuerier(deps)
+	blockClient, err := NewBlockClient(deps)
 	if err != nil {
 		return nil, err
 	}
 
-	blockClient, err := NewBlockClient(ctx, deps)
+	signer, err := NewSigner(signingKey)
 	if err != nil {
 		return nil, err
 	}
 
-	relayClient, err := NewRelayClient(ctx, deps)
+	shannonSDK := &ShannonSDK{
+		blockClient:   blockClient,
+		sessionClient: sessionClient,
+		accountClient: accountClient,
+		appClient:     appClient,
+		relayClient:   http.DefaultClient,
+		signer:        signer,
+	}
+
+	return shannonSDK, nil
+}
+
+// SendRelay builds a relay request from the given requestBz, signs it with the
+// application address, then sends it to the given endpoint.
+func (shannonSDK *ShannonSDK) SendRelay(
+	ctx context.Context,
+	appAddress string,
+	endpoint sdk.Endpoint,
+	requestBz []byte,
+) (*types.RelayResponse, error) {
+	relayRequest, err := sdk.BuildRelayRequest(endpoint, requestBz)
 	if err != nil {
 		return nil, err
 	}
 
-	signer, err := NewSigner(signingKey)
+	application, err := shannonSDK.appClient.GetApplication(ctx, appAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	appRing := sdk.ApplicationRing{
+		PublicKeyFetcher: shannonSDK.accountClient,
+		Application:      application,
+	}
+
+	shannonSDK.signer.Sign(ctx, relayRequest, appRing)
+
+	relayRequestBz, err := relayRequest.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := shannonSDK.relayClient.Post(
+		endpoint.Endpoint().Url,
+		"application/json",
+		bytes.NewReader(relayRequestBz),
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	responseBodyBz, err := io.ReadAll(response.Body)
 	if err != nil {
 		return nil, err
 	}
 
-	return sdk.NewShannonSDK(
-		applicationClient,
-		sessionClient,
-		accountClient,
-		sharedParamsClient,
-		blockClient,
-		relayClient,
-		signer,
+	return sdk.ValidateRelayResponse(
+		ctx,
+		endpoint.Supplier(),
+		responseBodyBz,
+		shannonSDK.accountClient,
 	)
 }
+
+// GetSessionSupplierEndpoints returns the current session's supplier endpoints
+// for the given appAddress and serviceId.
+func (shannonSDK *ShannonSDK) GetSessionSupplierEndpoints(
+	ctx context.Context,
+	appAddress, serviceId string,
+) (*sdk.FilteredSession, error) {
+	currentHeight := shannonSDK.blockClient.LastBlock(ctx).Height()
+	session, err := shannonSDK.sessionClient.GetSession(ctx, appAddress, serviceId, currentHeight)
+	if err != nil {
+		return nil, err
+	}
+
+	filteredSession := &sdk.FilteredSession{
+		Session: session,
+	}
+
+	return filteredSession, nil
```

Comment:
`FilteredSession` contains the session info with the corresponding filtered supplier endpoints (matching the service id). It can receive filter functions to further filter out suppliers or return all the (initially filtered) endpoints.

Ultimately it selects a single endpoint to send the relay request to.